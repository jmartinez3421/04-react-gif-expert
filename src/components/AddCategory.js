import React, {useState} from 'react';
import PropTypes from 'prop-types';

const AddCategory = ({setCategories}) => {

    const [inputValue, setInputValue] = useState('')

    const handleInputChange = (e) => {
        //En el atributo target.value del evento se encuentra el valor del input
        const nuevaCategoria = e.target.value;
        console.log('handleInputChange lanzado');

        //Actualizamos el useState con el valor introducido por el usuario
        setInputValue(nuevaCategoria);
    }

    const handleSubmit = (e) => {
        //Con el método preventDefault evitamos que el navegador se recargue al realizar el submit
        e.preventDefault();

        if( inputValue.trim().length >= 2){
            setCategories( categories => [inputValue, ...categories]);
            setInputValue('');
        }

        
    }

    return (
        <form onSubmit={handleSubmit}>
            {/* Cada vez que haya un cambio en el input se lanzará el evento onChange, el value del input será el 
            valor del useState */}
            <p style={{display:'none'}}>{inputValue}</p>
            <input 
                type="text" 
                value={inputValue} 
                onChange={handleInputChange}
                placeholder='¿Qué gifs quieres buscar?'
            />
        </form>
    )

}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}

export default AddCategory;