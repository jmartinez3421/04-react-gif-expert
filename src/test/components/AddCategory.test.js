import "@testing-library/jest-dom";
import { shallow } from 'enzyme';
import React from 'react';
import AddCategory from '../../components/AddCategory';

describe('Pruebas sobre el componente <AddCategory />', () => {
    const setCategories = jest.fn();
    let wrapper = shallow(<AddCategory setCategories={setCategories} />);

    beforeEach(() => {
        //Ejecutamos esta función para limpiar todos los valores antes de cada test
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategories={setCategories} />);
    })
    
    test('debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('debe llamar a handleInputChange cuando haya un cambio en el input', () => {
        const input = wrapper.find('input');
        const value = 'Hola mundo';

        input.simulate('change', { target: {value: value}});

        const pInputValue = wrapper.find('p');
        expect(pInputValue.text().trim()).toBe(value);
    });

    test('no debe de postear la información on submit si está vacío', () => {
        const form = wrapper.find('form');
        form.simulate('submit', { preventDefault(){} });

        expect(setCategories).not.toHaveBeenCalled();
    });

    test('debe llamar el setCategories y limpiar la caja de texto', () => {

        wrapper.find('input').simulate('change', {target: {value:'Simulo input change'}});
        wrapper.find('form').simulate('submit', {preventDefault(){}});

        expect(setCategories).toHaveBeenCalled();
        expect(wrapper.find('input').prop('value')).toBe('');
    })
})