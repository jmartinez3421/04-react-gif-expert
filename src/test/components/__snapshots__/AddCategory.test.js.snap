// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre el componente <AddCategory /> debe de mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <p
    style={
      Object {
        "display": "none",
      }
    }
  />
  <input
    onChange={[Function]}
    placeholder="¿Qué gifs quieres buscar?"
    type="text"
    value=""
  />
</form>
`;
